File: worker/src/byok-negotiation.ts
// BYOK price negotiation. The user argues their case in a real conversation
// with the model; the model may only *suggest* which of the server-defined
// concessions applies. The price itself is computed here, from
// ./byok-pricing, and written to D1 as an auditable record together with the
// conversation that produced it. No price value is ever read from a request
// body or from model output.

import { Hono } from "hono";
import {
  aiGatewayRoute,
  boundedJson,
  validatePinnedEndpoint,
  xiaomiCompletionEndpoint,
} from "./assistant";
import {
  type Concession,
  type PriceBand,
  concessionFor,
  formatPrice,
  normalizeGrants,
  priceBand,
  priceForGrants,
} from "./byok-pricing";
import { modelForTier } from "./model-tiers";
import { consumeRateLimit } from "./rate-limit";
import type { AppEnv, Bindings } from "./types";

const byok = new Hono<AppEnv>();

const maximumBodyBytes = 8 * 1024;
const maximumMessageCharacters = 600;
const maximumTranscriptEntries = 64;
const upstreamTimeoutMs = 20_000;
const sessionStartLimit = { limit: 3, windowMs: 24 * 3600_000 };
const messageLimit = { limit: 24, windowMs: 3600_000 };

export type TranscriptEntry = { role: "user" | "omi"; content: string };

type SessionRow = {
  id: string;
  uid: string;
  status: string;
  turns: number;
  grants: string;
  transcript: string;
};

export type AgreedPrice = {
  priceCents: number;
  outcome: "negotiated" | "standard";
  agreedAt: number;
};

const parseJsonArray = (value: unknown): unknown[] => {
  if (typeof value !== "string") return [];
  try {
    const parsed = JSON.parse(value) as unknown;
    return Array.isArray(parsed) ? parsed : [];
  } catch {
    return [];
  }
};

const parseTranscript = (value: unknown): TranscriptEntry[] =>
  parseJsonArray(value).flatMap((entry) => {
    if (entry === null || typeof entry !== "object" || Array.isArray(entry))
      return [];
    const record = entry as Record<string, unknown>;
    return (record.role === "user" || record.role === "omi") &&
      typeof record.content === "string"
      ? [{ role: record.role, content: record.content }]
      : [];
  });

// The recorded price for a user, or null when they have never settled one.
// Read straight from the audit record: the agreement row is the authority,
// not anything the client remembers.
export const agreedByokPrice = async (
  env: Bindings,
  uid: string,
): Promise<AgreedPrice | null> => {
  const row = await env.DB.prepare(
    "SELECT price_cents, outcome, agreed_at FROM byok_price_agreements WHERE uid = ?1",
  )
    .bind(uid)
    .first<{ price_cents: number; outcome: string; agreed_at: number }>();
  if (!row) return null;
  const band = priceBand(env);
  return {
    // Clamp on read as well as on write: a row written under an older, wider
    // band can never undercut the band in force today.
    priceCents: Math.min(
      band.standardCents,
      Math.max(band.floorCents, Number(row.price_cents)),
    ),
    outcome: row.outcome === "negotiated" ? "negotiated" : "standard",
    agreedAt: Number(row.agreed_at),
  };
};

const planPayload = (
  band: PriceBand,
  agreement: AgreedPrice | null,
  now: number,
) => ({
  standardPriceCents: band.standardCents,
  floorPriceCents: band.floorCents,
  priceCents: agreement?.priceCents ?? band.standardCents,
  outcome: agreement?.outcome ?? null,
  agreedAt: agreement?.agreedAt ?? null,
  negotiable: agreement === null || now >= agreement.agreedAt + band.cooldownMs,
  renegotiableAt:
    agreement === null ? null : agreement.agreedAt + band.cooldownMs,
});

const upsertAgreement = async (
  env: Bindings,
  uid: string,
  values: {
    sessionId: string | null;
    outcome: "negotiated" | "standard";
    priceCents: number;
    band: PriceBand;
    grants: readonly string[];
    transcript: TranscriptEntry[];
  },
  now: number,
): Promise<void> => {
  await env.DB.prepare(
    `INSERT INTO byok_price_agreements
       (uid, session_id, outcome, price_cents, standard_price_cents, floor_price_cents,
        grants, transcript, agreed_at, created_at, updated_at)
     VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?7, ?8, ?9, ?9, ?9)
     ON CONFLICT(uid) DO UPDATE SET
       session_id = excluded.session_id,
       outcome = excluded.outcome,
       price_cents = excluded.price_cents,
       standard_price_cents = excluded.standard_price_cents,
       floor_price_cents = excluded.floor_price_cents,
       grants = excluded.grants,
       transcript = excluded.transcript,
       agreed_at = excluded.agreed_at,
       updated_at = excluded.updated_at`,
  )
    .bind(
      uid,
      values.sessionId,
      values.outcome,
      values.priceCents,
      values.band.standardCents,
      values.band.floorCents,
      JSON.stringify(values.grants),
      JSON.stringify(values.transcript),
      now,
    )
    .run();
};

const systemPrompt = (band: PriceBand, granted: readonly string[]): string => {
  const available = band.concessions.filter(
    (concession) => !granted.includes(concession.code),
  );
  return [
    "You are Omi, negotiating your own subscription price with a user who has",
    "just connected their own AI provider key. Be warm, brief (two sentences",
    "at most), and honest. Never invent urgency, deadlines or scarcity.",
    "",
    "You do not set prices. You may only suggest at most one concession per",
    "reply, chosen from this list, when the user has genuinely made that case:",
    available.length === 0
      ? "(none left; you have nothing further to offer)"
      : available
          .map((concession) => `- ${concession.code}: ${concession.label}`)
          .join("\n"),
    "",
    "Never state a number, a price or a percentage; the app shows the price.",
    'Reply with JSON only: {"reply": string, "concession": string or null}.',
  ].join("\n");
};

type Suggestion = { reply: string; concession: Concession | null };

const parseSuggestion = (
  band: PriceBand,
  granted: readonly string[],
  raw: string,
): Suggestion | null => {
  const start = raw.indexOf("{");
  const end = raw.lastIndexOf("}");
  if (start < 0 || end <= start) return null;
  let parsed: unknown;
  try {
    parsed = JSON.parse(raw.slice(start, end + 1)) as unknown;
  } catch {
    return null;
  }
  if (parsed === null || typeof parsed !== "object" || Array.isArray(parsed))
    return null;
  const value = parsed as Record<string, unknown>;
  if (typeof value.reply !== "string" || value.reply.trim().length === 0)
    return null;
  const concession = concessionFor(band, value.concession);
  return {
    reply: value.reply.trim().slice(0, maximumMessageCharacters),
    // A concession already granted in this session cannot be granted twice,
    // whatever the model repeats.
    concession:
      concession && !granted.includes(concession.code) ? concession : null,
  };
};

// The model is told not to quote numbers, but "told not to" is not a control.
// Any currency or percentage figure that survives is replaced with the price
// the server computed, so the text can never disagree with the record.
export const sanitizeReply = (reply: string, priceCents: number): string =>
  reply
    .replace(/\$\s?\d+(?:[.,]\d+)?/g, formatPrice(priceCents))
    .replace(/\d+(?:\.\d+)?\s?%/g, "a bit");

const callModel = async (
  env: Bindings,
  band: PriceBand,
  granted: readonly string[],
  transcript: readonly TranscriptEntry[],
): Promise<string | null> => {
  const endpoint = env.MIMO_CHAT_COMPLETIONS_URL;
  const secret = env.MIMO_API_KEY;
  if (!endpoint || !secret) return null;
  const endpointUrl = validatePinnedEndpoint(
    endpoint,
    xiaomiCompletionEndpoint,
    "token-plan-sgp.xiaomimimo.com",
  );
  if (!endpointUrl) return null;
  const gateway = aiGatewayRoute(env);
  try {
    const upstream = await fetch(gateway?.url ?? endpointUrl, {
      method: "POST",
      headers: {
        authorization: `Bearer ${secret}`,
        "content-type": "application/json",
        ...gateway?.headers,
      },
      signal: AbortSignal.timeout(upstreamTimeoutMs),
      body: JSON.stringify({
        model: modelForTier(env, "balanced"),
        stream: false,
        max_tokens: 400,
        temperature: 0.7,
        messages: [
          { role: "system", content: systemPrompt(band, granted) },
          ...transcript.map((entry) => ({
            role: entry.role === "user" ? "user" : "assistant",
            content: entry.content,
          })),
        ],
      }),
    });
    if (!upstream.ok) return null;
    const completion = (await upstream.json()) as {
      choices?: Array<{ message?: { content?: unknown } }>;
    };
    const content = completion.choices?.[0]?.message?.content;
    return typeof content === "string" ? content : null;
  } catch {
    return null;
  }
};

const loadSession = async (
  env: Bindings,
  uid: string,
  id: string,
): Promise<SessionRow | null> => {
  const row = await env.DB.prepare(
    "SELECT id, uid, status, turns, grants, transcript FROM byok_negotiation_sessions WHERE id = ?1 AND uid = ?2",
  )
    .bind(id, uid)
    .first<SessionRow>();
  return row ?? null;
};

byok.get("/plan", async (context) => {
  const band = priceBand(context.env);
  const agreement = await agreedByokPrice(context.env, context.get("auth").uid);
  return context.json(planPayload(band, agreement, Date.now()));
});

// Taking the standard price is always available and is recorded like any
// other outcome, so skipping is a first-class path rather than a dead end.
byok.post("/plan/standard", async (context) => {
  const band = priceBand(context.env);
  const uid = context.get("auth").uid;
  const now = Date.now();
  await upsertAgreement(
    context.env,
    uid,
    {
      sessionId: null,
      outcome: "standard",
      priceCents: band.standardCents,
      band,
      grants: [],
      transcript: [],
    },
    now,
  );
  return context.json(
    planPayload(band, await agreedByokPrice(context.env, uid), now),
    201,
  );
});

byok.post("/negotiation", async (context) => {
  const band = priceBand(context.env);
  const uid = context.get("auth").uid;
  const now = Date.now();
  const agreement = await agreedByokPrice(context.env, uid);
  // Renegotiation is bounded twice: by the recorded agreement's cooldown and
  // by a rate limiter, so repeat attempts cannot be farmed for a lower price.
  if (agreement && now < agreement.agreedAt + band.cooldownMs)
    return context.json(
      { error: "Price already agreed", ...planPayload(band, agreement, now) },
      409,
    );
  const limit = await consumeRateLimit(
    context.env,
    `byok-negotiation-start:${uid}`,
    sessionStartLimit.limit,
    sessionStartLimit.windowMs,
  );
  if (!limit.allowed)
    return context.json({ error: "Too many negotiations" }, 429, {
      "retry-after": String(limit.retryAfter),
    });
  if (!context.env.MIMO_API_KEY || !context.env.MIMO_CHAT_COMPLETIONS_URL)
    return context.json({ error: "Negotiation unavailable" }, 503);
  const id = crypto.randomUUID();
  const opening: TranscriptEntry = {
    role: "omi",
    content:
      `Standard with your own key is ${formatPrice(band.standardCents)} a month. ` +
      "If that is not right for you, tell me why and I will see what I can do.",
  };
  await context.env.DB.prepare(
    `INSERT INTO byok_negotiation_sessions
       (id, uid, status, turns, standard_price_cents, floor_price_cents, price_cents,
        grants, transcript, created_at, updated_at)
     VALUES (?1, ?2, 'open', 0, ?3, ?4, ?3, '[]', ?5, ?6, ?6)`,
  )
    .bind(
      id,
      uid,
      band.standardCents,
      band.floorCents,
      JSON.stringify([opening]),
      now,
    )
    .run();
  return context.json(
    {
      sessionId: id,
      priceCents: band.standardCents,
      standardPriceCents: band.standardCents,
      turnsRemaining: band.maxTurns,
      transcript: [opening],
    },
    201,
  );
});

byok.post("/negotiation/:id/message", async (context) => {
  const band = priceBand(context.env);
  const uid = context.get("auth").uid;
  const session = await loadSession(context.env, uid, context.req.param("id"));
  if (!session) return context.json({ error: "Unknown negotiation" }, 404);
  if (session.status !== "open")
    return context.json({ error: "Negotiation closed" }, 409);
  const body = await boundedJson(context.req.raw, maximumBodyBytes);
  const message =
    typeof body?.message === "string" ? body.message.trim() : null;
  if (!message || message.length > maximumMessageCharacters)
    return context.json({ error: "Invalid request" }, 400);
  const limit = await consumeRateLimit(
    context.env,
    `byok-negotiation-message:${uid}`,
    messageLimit.limit,
    messageLimit.windowMs,
  );
  if (!limit.allowed)
    return context.json({ error: "Too many messages" }, 429, {
      "retry-after": String(limit.retryAfter),
    });
  if (session.turns >= band.maxTurns)
    return context.json({ error: "Negotiation closed" }, 409);
  const granted = normalizeGrants(band, parseJsonArray(session.grants));
  const transcript = parseTranscript(session.transcript).slice(
    -maximumTranscriptEntries,
  );
  transcript.push({ role: "user", content: message });
  const raw = await callModel(context.env, band, granted, transcript);
  const suggestion = raw === null ? null : parseSuggestion(band, granted, raw);
  if (!suggestion)
    return context.json({ error: "Negotiation unavailable" }, 502);
  const grants = suggestion.concession
    ? [...granted, suggestion.concession.code]
    : granted;
  const priceCents = priceForGrants(band, grants);
  const reply = sanitizeReply(suggestion.reply, priceCents);
  transcript.push({ role: "omi", content: reply });
  const turns = session.turns + 1;
  const now = Date.now();
  await context.env.DB.prepare(
    `UPDATE byok_negotiation_sessions
       SET turns = ?1, grants = ?2, transcript = ?3, price_cents = ?4, updated_at = ?5
     WHERE id = ?6 AND uid = ?7`,
  )
    .bind(
      turns,
      JSON.stringify(grants),
      JSON.stringify(transcript),
      priceCents,
      now,
      session.id,
      uid,
    )
    .run();
  return context.json({
    reply,
    priceCents,
    standardPriceCents: band.standardCents,
    turnsRemaining: Math.max(0, band.maxTurns - turns),
    conceded: suggestion.concession !== null,
  });
});

// Accepting recomputes the price from the stored grants rather than trusting
// anything in the request, so a replayed or edited accept settles at exactly
// the same figure the conversation earned.
byok.post("/negotiation/:id/accept", async (context) => {
  const band = priceBand(context.env);
  const uid = context.get("auth").uid;
  const session = await loadSession(context.env, uid, context.req.param("id"));
  if (!session) return context.json({ error: "Unknown negotiation" }, 404);
  const now = Date.now();
  if (session.status === "agreed")
    return context.json(
      planPayload(band, await agreedByokPrice(context.env, uid), now),
      200,
    );
  if (session.status !== "open")
    return context.json({ error: "Negotiation closed" }, 409);
  const grants = normalizeGrants(band, parseJsonArray(session.grants));
  const priceCents = priceForGrants(band, grants);
  await context.env.DB.prepare(
    "UPDATE byok_negotiation_sessions SET status = 'agreed', price_cents = ?1, updated_at = ?2 WHERE id = ?3 AND uid = ?4",
  )
    .bind(priceCents, now, session.id, uid)
    .run();
  await upsertAgreement(
    context.env,
    uid,
    {
      sessionId: session.id,
      outcome: priceCents < band.standardCents ? "negotiated" : "standard",
      priceCents,
      band,
      grants,
      transcript: parseTranscript(session.transcript),
    },
    now,
  );
  return context.json(
    planPayload(band, await agreedByokPrice(context.env, uid), now),
    201,
  );
});

export default byok;

File: worker-rs/src/byok_negotiation.rs
//! Pure parity port of the decision logic in `worker/src/byok-negotiation.ts`.
//!
//! The user argues their case in a real conversation with the model; the model
//! may only *suggest* which of the server-defined concessions applies. The
//! price itself is computed in [`crate::byok_pricing`] and written to D1 as an
//! auditable record together with the conversation that produced it. No price
//! value is ever read from a request body or from model output.

use serde_json::{json, Value};

use crate::byok_pricing::{concession_for, format_price, Concession, PriceBand};

pub const MAXIMUM_BODY_BYTES: usize = 8 * 1024;
pub const MAXIMUM_MESSAGE_CHARACTERS: usize = 600;
pub const MAXIMUM_TRANSCRIPT_ENTRIES: usize = 64;
pub const UPSTREAM_TIMEOUT_MS: i64 = 20_000;
/// `sessionStartLimit` / `messageLimit`.
pub const SESSION_START_LIMIT: i64 = 3;
pub const SESSION_START_WINDOW_MS: i64 = 24 * 3_600_000;
pub const MESSAGE_LIMIT: i64 = 24;
pub const MESSAGE_WINDOW_MS: i64 = 3_600_000;

#[derive(Clone, Debug, PartialEq)]
pub struct TranscriptEntry {
    pub role: String,
    pub content: String,
}

impl TranscriptEntry {
    pub fn to_value(&self) -> Value {
        json!({ "role": self.role, "content": self.content })
    }
}

#[derive(Clone, Debug, PartialEq)]
pub struct AgreedPrice {
    pub price_cents: i64,
    pub outcome: String,
    pub agreed_at: i64,
}

/// A JSON-encoded array column; anything else is an empty list.
pub fn parse_json_array(value: Option<&Value>) -> Vec<Value> {
    let Some(Value::String(raw)) = value else {
        return Vec::new();
    };
    match serde_json::from_str::<Value>(raw) {
        Ok(Value::Array(items)) => items,
        _ => Vec::new(),
    }
}

/// `parseTranscript` — only `user`/`omi` entries with string content survive.
pub fn parse_transcript(value: Option<&Value>) -> Vec<TranscriptEntry> {
    parse_json_array(value)
        .iter()
        .filter_map(|entry| {
            let record = entry.as_object()?;
            let role = record.get("role")?.as_str()?;
            let content = record.get("content")?.as_str()?;
            (role == "user" || role == "omi").then(|| TranscriptEntry {
                role: role.to_string(),
                content: content.to_string(),
            })
        })
        .collect()
}

/// Clamp a stored agreement into the band in force today: a row written under
/// an older, wider band can never undercut the current floor or ceiling.
pub fn clamp_agreement(
    band: &PriceBand,
    price_cents: i64,
    outcome: &str,
    agreed_at: i64,
) -> AgreedPrice {
    AgreedPrice {
        price_cents: band.standard_cents.min(band.floor_cents.max(price_cents)),
        outcome: if outcome == "negotiated" {
            "negotiated".to_string()
        } else {
            "standard".to_string()
        },
        agreed_at,
    }
}

/// `planPayload(band, agreement, now)`.
pub fn plan_payload(band: &PriceBand, agreement: Option<&AgreedPrice>, now: i64) -> Value {
    json!({
        "standardPriceCents": band.standard_cents,
        "floorPriceCents": band.floor_cents,
        "priceCents": agreement.map_or(band.standard_cents, |a| a.price_cents),
        "outcome": agreement.map(|a| a.outcome.clone()),
        "agreedAt": agreement.map(|a| a.agreed_at),
        "negotiable": agreement.is_none_or(|a| now >= a.agreed_at + band.cooldown_ms),
        "renegotiableAt": agreement.map(|a| a.agreed_at + band.cooldown_ms),
    })
}

/// The opening line of a fresh negotiation.
pub fn opening_entry(band: &PriceBand) -> TranscriptEntry {
    TranscriptEntry {
        role: "omi".to_string(),
        content: format!(
            "Standard with your own key is {} a month. If that is not right for you, tell me why and I will see what I can do.",
            format_price(band.standard_cents)
        ),
    }
}

/// `systemPrompt(band, granted)`.
pub fn system_prompt(band: &PriceBand, granted: &[String]) -> String {
    let available: Vec<&Concession> = band
        .concessions
        .iter()
        .filter(|concession| !granted.iter().any(|code| code == concession.code))
        .collect();
    let offer = if available.is_empty() {
        "(none left; you have nothing further to offer)".to_string()
    } else {
        available
            .iter()
            .map(|concession| format!("- {}: {}", concession.code, concession.label))
            .collect::<Vec<_>>()
            .join("\n")
    };
    [
        "You are Omi, negotiating your own subscription price with a user who has",
        "just connected their own AI provider key. Be warm, brief (two sentences",
        "at most), and honest. Never invent urgency, deadlines or scarcity.",
        "",
        "You do not set prices. You may only suggest at most one concession per",
        "reply, chosen from this list, when the user has genuinely made that case:",
        &offer,
        "",
        "Never state a number, a price or a percentage; the app shows the price.",
        r#"Reply with JSON only: {"reply": string, "concession": string or null}."#,
    ]
    .join("\n")
}

pub struct Suggestion {
    pub reply: String,
    pub concession: Option<Concession>,
}

/// `parseSuggestion` — the model's output is untrusted text. Only the reply
/// string and a *known, not-yet-granted* concession code survive.
pub fn parse_suggestion(band: &PriceBand, granted: &[String], raw: &str) -> Option<Suggestion> {
    let start = raw.find('{')?;
    let end = raw.rfind('}')?;
    if end <= start {
        return None;
    }
    let parsed = serde_json::from_str::<Value>(&raw[start..=end]).ok()?;
    let record = parsed.as_object()?;
    let reply = record.get("reply")?.as_str()?;
    if reply.trim().is_empty() {
        return None;
    }
    let concession = concession_for(band, record.get("concession").and_then(Value::as_str))
        // A concession already granted in this session cannot be granted
        // twice, whatever the model repeats.
        .filter(|concession| !granted.iter().any(|code| code == concession.code))
        .cloned();
    Some(Suggestion {
        reply: reply
            .trim()
            .chars()
            .take(MAXIMUM_MESSAGE_CHARACTERS)
            .collect(),
        concession,
    })
}

/// The model is told not to quote numbers, but "told not to" is not a control.
/// Any currency or percentage figure that survives is replaced with the price
/// the server computed, so the text can never disagree with the record.
///
/// Mirrors `reply.replace(/\$\s?\d+(?:[.,]\d+)?/g, formatPrice(price))
///              .replace(/\d+(?:\.\d+)?\s?%/g, "a bit")`.
pub fn sanitize_reply(reply: &str, price_cents: i64) -> String {
    replace_percentages(&replace_currency(reply, &format_price(price_cents)))
}

fn replace_currency(input: &str, replacement: &str) -> String {
    let chars: Vec<char> = input.chars().collect();
    let mut out = String::with_capacity(input.len());
    let mut index = 0;
    while index < chars.len() {
        if chars[index] == '$' {
            if let Some(end) = currency_match(&chars, index) {
                out.push_str(replacement);
                index = end;
                continue;
            }
        }
        out.push(chars[index]);
        index += 1;
    }
    out
}

/// `\$\s?\d+(?:[.,]\d+)?` anchored at `start`; returns the end index.
fn currency_match(chars: &[char], start: usize) -> Option<usize> {
    let mut index = start + 1;
    if chars.get(index).is_some_and(|c| c.is_whitespace()) {
        index += 1;
    }
    let digits_start = index;
    while chars.get(index).is_some_and(char::is_ascii_digit) {
        index += 1;
    }
    if index == digits_start {
        return None;
    }
    if matches!(chars.get(index), Some('.') | Some(','))
        && chars.get(index + 1).is_some_and(char::is_ascii_digit)
    {
        index += 1;
        while chars.get(index).is_some_and(char::is_ascii_digit) {
            index += 1;
        }
    }
    Some(index)
}

fn replace_percentages(input: &str) -> String {
    let chars: Vec<char> = input.chars().collect();
    let mut out = String::with_capacity(input.len());
    let mut index = 0;
    while index < chars.len() {
        if let Some(end) = percentage_match(&chars, index) {
            out.push_str("a bit");
            index = end;
            continue;
        }
        out.push(chars[index]);
        index += 1;
    }
    out
}

/// `\d+(?:\.\d+)?\s?%` anchored at `start`; returns the end index.
fn percentage_match(chars: &[char], start: usize) -> Option<usize> {
    let mut index = start;
    while chars.get(index).is_some_and(char::is_ascii_digit) {
        index += 1;
    }
    if index == start {
        return None;
    }
    if chars.get(index) == Some(&'.') && chars.get(index + 1).is_some_and(char::is_ascii_digit) {
        index += 1;
        while chars.get(index).is_some_and(char::is_ascii_digit) {
            index += 1;
        }
    }
    if chars.get(index).is_some_and(|c| c.is_whitespace()) {
        index += 1;
    }
    (chars.get(index) == Some(&'%')).then_some(index + 1)
}

/// `POST /v1/byok/negotiation/:id/message` body validation.
pub fn validate_message(body: Option<&Value>) -> Option<String> {
    let raw = body?.get("message")?.as_str()?;
    let message = raw.trim();
    (!message.is_empty() && message.chars().count() <= MAXIMUM_MESSAGE_CHARACTERS)
        .then(|| message.to_string())
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::byok_pricing::{price_band, price_for_grants};

    fn band() -> PriceBand {
        price_band(|_| None)
    }

    #[test]
    fn plan_payload_reflects_the_cooldown() {
        let band = band();
        let open = plan_payload(&band, None, 1_000);
        assert_eq!(open["priceCents"], json!(1_200));
        assert_eq!(open["negotiable"], json!(true));
        assert_eq!(open["outcome"], Value::Null);
        assert_eq!(open["renegotiableAt"], Value::Null);

        let agreement = AgreedPrice {
            price_cents: 900,
            outcome: "negotiated".into(),
            agreed_at: 1_000,
        };
        let fresh = plan_payload(&band, Some(&agreement), 2_000);
        assert_eq!(fresh["priceCents"], json!(900));
        assert_eq!(fresh["negotiable"], json!(false));
        assert_eq!(fresh["renegotiableAt"], json!(1_000 + band.cooldown_ms));
        let later = plan_payload(&band, Some(&agreement), 1_000 + band.cooldown_ms);
        assert_eq!(later["negotiable"], json!(true));
    }

    #[test]
    fn a_stored_row_is_clamped_into_the_band_on_read() {
        let band = band();
        assert_eq!(clamp_agreement(&band, 1, "negotiated", 0).price_cents, 700);
        assert_eq!(
            clamp_agreement(&band, 99_999, "standard", 0).price_cents,
            1_200
        );
        assert_eq!(clamp_agreement(&band, 900, "weird", 0).outcome, "standard");
    }

    #[test]
    fn transcript_parsing_drops_foreign_roles_and_shapes() {
        let raw = json!(
            r#"[{"role":"user","content":"hi"},{"role":"system","content":"x"},{"role":"omi","content":"there"},{"role":"omi"},null,7]"#
        );
        assert_eq!(
            parse_transcript(Some(&raw)),
            vec![
                TranscriptEntry {
                    role: "user".into(),
                    content: "hi".into()
                },
                TranscriptEntry {
                    role: "omi".into(),
                    content: "there".into()
                },
            ]
        );
        assert!(parse_transcript(Some(&json!("nonsense"))).is_empty());
        assert!(parse_transcript(None).is_empty());
    }

    #[test]
    fn grants_a_concession_the_model_suggests_once() {
        let band = band();
        let raw = r#"Sure! {"reply":"That is fair.","concession":"student"}"#;
        let first = parse_suggestion(&band, &[], raw).unwrap();
        assert_eq!(first.concession.as_ref().unwrap().code, "student");
        let granted = vec!["student".to_string()];
        let second = parse_suggestion(&band, &granted, raw).unwrap();
        assert!(second.concession.is_none());
    }

    #[test]
    fn a_manipulated_model_reply_cannot_invent_a_lever_or_a_price() {
        let band = band();
        for raw in [
            r#"{"reply":"ok","concession":"free_forever"}"#,
            r#"{"reply":"ok","concession":{"code":"student","centsOff":1200}}"#,
            r#"{"reply":"ok","concession":99}"#,
        ] {
            let suggestion = parse_suggestion(&band, &[], raw).unwrap();
            assert!(suggestion.concession.is_none(), "should ignore {raw}");
            assert_eq!(price_for_grants(&band, &[]), band.standard_cents);
        }
        for raw in [
            "",
            "no json here",
            r#"{"reply":"   "}"#,
            r#"{"reply":7}"#,
            "}{",
        ] {
            assert!(
                parse_suggestion(&band, &[], raw).is_none(),
                "should reject {raw:?}"
            );
        }
    }

    #[test]
    fn a_reply_is_trimmed_and_bounded() {
        let band = band();
        let long = "a".repeat(2_000);
        let raw = format!(r#"{{"reply":"  {long}  "}}"#);
        let suggestion = parse_suggestion(&band, &[], &raw).unwrap();
        assert_eq!(suggestion.reply.chars().count(), MAXIMUM_MESSAGE_CHARACTERS);
    }

    #[test]
    fn sanitize_replaces_every_figure_the_model_slipped_in() {
        assert_eq!(
            sanitize_reply("I can do $5 or even $ 4.50 for you.", 900),
            "I can do $9.00 or even $9.00 for you."
        );
        assert_eq!(
            sanitize_reply("A 30% discount, 7.5 % off.", 900),
            "A a bit discount, a bit off."
        );
        assert_eq!(sanitize_reply("$1,299", 700), "$7.00");
        // Bare numbers that are neither prices nor percentages are untouched.
        assert_eq!(sanitize_reply("You get 3 seats.", 900), "You get 3 seats.");
        assert_eq!(sanitize_reply("", 900), "");
    }

    #[test]
    fn system_prompt_only_offers_ungranted_concessions() {
        let band = band();
        let prompt = system_prompt(&band, &["student".to_string()]);
        assert!(!prompt.contains("- student:"));
        assert!(prompt.contains("- case_study:"));
        let exhausted: Vec<String> = band
            .concessions
            .iter()
            .map(|c| c.code.to_string())
            .collect();
        assert!(system_prompt(&band, &exhausted).contains("(none left"));
    }

    #[test]
    fn message_validation_bounds_the_input() {
        assert_eq!(
            validate_message(Some(&json!({ "message": "  I am a student " }))).as_deref(),
            Some("I am a student")
        );
        for body in [
            json!({ "message": "   " }),
            json!({ "message": 7 }),
            json!({}),
            json!({ "message": "x".repeat(601) }),
        ] {
            assert!(
                validate_message(Some(&body)).is_none(),
                "should reject {body}"
            );
        }
        assert!(validate_message(None).is_none());
    }

    #[test]
    fn the_opening_line_quotes_the_standard_price() {
        let opening = opening_entry(&band());
        assert_eq!(opening.role, "omi");
        assert!(opening.content.contains("$12.00"));
    }
}
